Drag external objects onto layout to make tabs

edemaine opened the pull request:
This PR adds a configuration option `onExternalDrag` to allow the layout to handle drags from outside the layout (e.g. files or links or even draggable objects from other parts of the app outside the layout) that cause the creation of tabs.  In [my app](https://github.com/edemaine/comingle), I want to be able to drag URLs from e.g. another browser window in order to create a tab that iframes that URL.

Most of the changes are local to DragDrop, which has extra event handlers for the many various drag/drop events (and the usual [counter workaround](https://www.smashingmagazine.com/2020/02/html-drag-drop-api-react/) to handle dragging entering/leaving various elements).  The main other change is that Layout needs to listen to `dragenter` as yet another way to start a new drag operation.

I hope you like it!  Let me know if you want it tweaked in any way.

(This is the main feature addition that led me down the path to the other PR and issue.)

edemaine commented:
I realize that this needs a little more work: the `onDrop` callback needs to be given the `drop` event; otherwise, it's impossible to receive the dropped `dataTransfer`.  (I thought it could be accessed during the `dragenter` event passed to `onExternalDrag`, but it turns out that `dataTransfer` is in protected mode so you can see the data types but not the data.)

edemaine commented:
OK, this is now functional, though I wonder if there's an easier way to configure the tab after a successful drop.  Here is how I use it currently:

```js
onExternalDrag={(e) => {
  var json;
  // Check for supported content type
  if (e.dataTransfer.types.indexOf('text/plain') < 0) return;
  // Set dropEffect (icon)
  e.dataTransfer.dropEffect = 'link'
  return {
    dragText: "Drag New Tab",
    json: json = {
      id: 'drop',
      type: 'tab',
      name: 'New Tab',
      component: 'TabNew',
      enableRename: false,
    },
    onDrop: (drop) => {
      if (!drop) return;  // aborted drag
      var tab = model.getNodeById('drop');
      if (!tab) return;  // shouldn't happen, but the usual way to check for aborted drag
      var text = drop.dataTransfer.getData('text/plain');
      var url = ... parse text as desired ...
      // Replace place-holder tab with actual tab (to avoid id duplication)
      if (url) {
        delete json.id;
        json.config = {url}
        model.doAction(FlexLayout.Actions.addNode(json,
          tab.getParent().getId(), FlexLayout.DockLocation.CENTER, -1));
      }
      model.doAction FlexLayout.Actions.deleteTab('drop');
    }
  }
}
```

I think it would be nicer if the `onDrop` handler got passed in (probably as first argument) the added tab, so that I could use auto-generated IDs.  This would help with uses of `addTabWithDragAndDrop` as well (it suffers from the same problem).  To fix this, it seems like we need to add a return value to `doAction` in the case of `addNode` &mdash; would that be OK?

nealus commented:
Yes, I think adding the tab as the first arg of onDrop and returning it from the doAction for addNode makes sense

edemaine commented:
OK, I've implemented this change (and rebased onto master).  The usage code now looks like this:

```js
onExternalDrag={(e) => {
  // Check for supported content type
  if (e.dataTransfer.types.indexOf('text/plain') < 0) return;
  // Set dropEffect (icon)
  e.dataTransfer.dropEffect = 'link';
  return {
    dragText: "Drag New Tab",
    json: {
      type: 'tab',
      name: 'New Tab',
      component: 'TabNew',
      enableRename: false,
    },
    onDrop: (tab, drop) => {
      if (!tab || !drop) return;  // aborted drag
      var text = drop.dataTransfer.getData('text/plain');
      var url = ... parse text as desired ...
      model.doAction(FlexLayout.Actions.updateNodeAttributes(tab.getId(), {config: {url}});
    }
  }
}}
```

It seems to be working well!

nealus commented:
Great, I think you should also be able to do your example like this:

```
onExternalDrag={(e) => {
  // Check for supported content type
  if (e.dataTransfer.types.indexOf('text/plain') < 0) return;
  // Set dropEffect (icon)
  e.dataTransfer.dropEffect = 'link';
  var text = e.dataTransfer.getData('text/plain');
  var url = ... parse text as desired ...
  return {
    dragText: "Drag New Tab",
    json: {
      type: 'tab',
      name: 'New Tab',
      component: 'TabNew',
      enableRename: false,
      config: {url}
    },
    onDrop: (tab, drop) => {
    }
  }
}}
```

assuming you can get the e.dataTransfer.getData at this point

edemaine commented:
No, unfortunately, that's not possible.  `onExternalDrag` is called from the `dragenter` event (the first event that you get from a drag that occurs from an external window), and at that point, the `dataTransfer` object is in "protected" mode, so you can read the types but not the content.  Only once we get a `drop` event can we read the drop contents.  See e.g. [this thread](https://stackoverflow.com/questions/40940288/drag-datatransfer-data-unavailable-in-ondragover-event).

nealus commented:
OK, good to know, thanks

nealus commented:
Merged and added simple example to demo (https://rawgit.com/caplin/FlexLayout/demos/demos/v0.5/demo/index.html)

nealus commented:
Added a small fix ref dragdrop glasspane pointer events, these are needed for dragging over iframes, but need to be disabled for external drag/drop

nealus commented:
In Chrome, in a tab with an iframe, if you drag a link, just a small movement is needed, it completely breaks external 
drag and drop. In fact it will not even come back after a page reload.

The same thing works ok in firefox.

You can see it in the demo: https://rawgit.com/caplin/FlexLayout/demos/demos/v0.5/demo/index.html
just try moving the 'Main Page' link on the wikipedia tab, afterwards external drag and drop stops working.

I've tried moving the onDragEnter from in the <Layout to an addEventListener (and re-adding it) but no luck so far.

edemaine commented:
Cool idea to add it to the demo, thanks!

I took a look at the current state of the demo, as well as my original branch, on Chrome (Windows). It looks like:
* drags from other browser windows (actual external drag) are working fine
* dragging from within an iframe in the FlexLayout does not work (odd, I thought I tested this...)
* dragging onto an iframe in the FlexLayout does not work (but if you aim for a tab button, say, it works)

Now that I understand the purpose of the glass is to deal with iframes, I'll look into this more...

nealus commented:
If you 'reload from file' on the Default view of the demo the Wikipedia page should show (does for me, could be a browser caching issue).

I'm seeing the problem with the external drag breaking for the browser window on Chrome for Mac Version 89.0.4389.114 (Official Build) (x86_64)), I think this must be Chrome (on Mac?) bug, because even after a reload, that tab will never do external drag and drop again.

edemaine commented:
Indeed, reload from file does it.

I am on Windows, so perhaps that's the difference... Might also be related to [the security feature](https://bugs.chromium.org/p/chromium/issues/detail?id=59081).

nealus commented:
Just tried on windows, and external drag and drop doesn't break after a drag in an iframe, so yes it's a Mac issue, I've tried on both Catalina and Mojave

Jaredanwolfgang commented:
Hi, I am currently working on a project using flexlayout-react and encounter relevant issue regarding the topic. I am trying to use FlexLayout-react to realize an IDE-like interaction panel where you can either edit file (The editor I use is `monaco-editor`) or view pdf (The package I use is `react-pdf` ) inside each tab. I am self defining the function of `handleExternalDrag` for the `onExternalDrag` so that I can drag a file in the file tree (the file bar on the left) inside the panel. Here is the implementation I use:

```typescript
const handleExternalDrag = (event: React.DragEvent<HTMLElement>) => {
    if (event.dataTransfer.types.indexOf('application/json') < 0) return;
    event.dataTransfer.dropEffect = 'link';
    return {
      dragText: `Drag New File`,
      json: {
        type: "tab",
        name: "New File",
        component: "editor", 
        config: { filePath: "New File", url: undefined },
        language: "plaintext",
      },
      onDrop: (node?: Node, event?: React.DragEvent<HTMLElement>) => {
        if (!node || !event) return; // Aborted drag
        const uri = event.dataTransfer.getData("text/plain");
        const language = getLanguageFromFileName(uri);
        const isPDF = language === 'pdf';
        model.doAction(Actions.updateNodeAttributes(node.getId(), {
          name: uri.split('/').pop() || uri,
          component: isPDF ? "pdf" : "editor",
          config: { filePath: uri, url: isPDF ? uri : undefined },
          language
        }));
      }
    };
  };
```

This is nearly the same as the example above. The "text/plain" is set when the file component is `onDragBegin`. 

The question is, will the implementation of the `onDrop()` function will affect the other components of the Layout. It seems like for each "drop", the whole layout will be updated. And since I try to render a `pdf` within the layout. The `pdf` fails to re-render each time I trigger the `onDrop()` function.(Only when I try to move the tab within the panel.) Is there anyway that the implementation of `onDrop` function is isolated and will not affect the inside `drop` event? (I assume this is the problem. Probably I get this wrong. Correct me if I am wrong.) 

<img width="784" alt="Screenshot 2025-04-03 at 22 24 25" src="https://github.com/user-attachments/assets/b963842e-2697-420b-8982-b6e00344c4e7" />

When I try to move it it becomes: (If I do not implement the onExternalDrag, the Drag within the Layout will not trigger this re-render?)

<img width="803" alt="Screenshot 2025-04-03 at 22 22 32" src="https://github.com/user-attachments/assets/be01588a-0b8d-4543-80e2-60527ff21730" />


Thank you again for the help!

(I am using React + Next.js.)
